### DISCUSSION TITLE
Support for ICAP protocol?

### DISCUSSION THREAD

**Discussion by Daniel-Nashed** [curl/curl#7694] [Ideas]
Hi,

I didn't find any information about support for [ICAP](https://datatracker.ietf.org/doc/html/rfc3507).
And I have seen it has been first asked around 2007. Before writing my own integration or looking into some other project, I would like to double check if I might have overlooked something on the Curl side.

The protocol is very HTTP like with some very specific optimizations in the flow mainly introduced for proxies sending their data stream to an antivirus server.

In my case I am looking for a simple interface to use ICAP for attachment scanning.

Before reinventing the wheel I would see if LibCurl has some kind of support for it.
Or at least if I could use LibCurl for the base communication back-end and add some protocol specific additions on top if it.

Is there any info or did anyone look into this before?

Thanks in advance

Daniel

**Comment by bagder** [comment]
I don't think anyone has looked into it closer and I can't recall anyone particularly asking for it either after that mailing list post in 2007. I personally don't really know what it takes to support it and what's needed to make it "attractive". It being basically HTTP sounds like a decent start though, but also possibly tricky if you need to limit the http subset.

**Reply by Daniel-Nashed** [reply]
Thanks for your prompt reply ! And sorry for the delay.

I just wanted to make sure I am not overlooking something that is already there.
The protocol is similar to HTTP but not exactly the same. And it could also have been that someone combined libcurl functionality to make it work. I have to look closer into the protocol and there is the [c-icap](http://c-icap.sourceforge.net/) which could be used.
But this means adding another lib and more complexity with unknown risks and challenges. 

Having a way to use libcurl would have been great but I don't see that this should be looked into! There are other features that would be much higher on people's wish list and not many people would benefit from supporting ICAP.

Thanks again for all your great work!

Daniel

**Comment by apollo7321** [comment]
@bagder since there is no official support by curl itself, I was wondering if there is a way to use libcurl to add the additional headers above the http headers itself?

According to the RFC there are ICAP headers with a blank line between the ICAP and the HTTP header like this:

```
   REQMOD icap://icap-server.net/server?arg=87 ICAP/1.0
   Host: icap-server.net
   Encapsulated: req-hdr=0, null-body=170

   GET / HTTP/1.1
   Host: www.origin-server.com
   Accept: text/html, text/plain
   Accept-Encoding: compress
   Cookie: ff39fk3jur@4ii0e02i
   If-None-Match: "xyzzy", "r2d2xxxx"
```

Is there any way to achieve this with libcurl? So far I am not sure.
